test/common/quic/envoy_quic_h3_fuzz_test.cc

```diff
@@ -0,0 +1,768 @@
+    session->setHttp3Options(http3_options_);
+    session->setCodecStats(http3_stats_);
+    session->setHttpConnectionCallbacks(http_connection_callbacks_);
+    session->setMaxIncomingHeadersCount(100);
+    session->set_max_inbound_header_list_size(64 * 1024u);
+    setQuicConfigWithDefaultValues(session->config());
+    session->OnConfigNegotiated();
+    return session;
+  }
+
+private:
+  NiceMock<Network::MockListenerConfig> mock_listener_config_;
+  Network::ListenerConfig* listener_config_{};
+  QuicDispatcherStats quic_stats_;
+  Event::Dispatcher& dispatcher_;
+
+  envoy::config::core::v3::Http3ProtocolOptions http3_options_;
+  quic::ParsedQuicVersion quic_version_;
+  QuicPacketizer packetizer_;
+  quic::QuicCryptoServerConfig crypto_config_;
+  NiceMock<quic::test::MockQuicCryptoServerStreamHelper> crypto_stream_helper_;
+  Network::Address::InstanceConstSharedPtr peer_addr_;
+  Network::Address::InstanceConstSharedPtr self_addr_;
+  quic::QuicSocketAddress cli_addr_;
+  quic::QuicSocketAddress srv_addr_;
+  QuicStatNames quic_stat_names_;
+  Http::Http3::CodecStats http3_stats_;
+  QuicConnectionStats connection_stats_;
+
+  quic::QuicConfig quic_config_;
+  quic::DeterministicConnectionIdGenerator generator_{quic::kQuicDefaultConnectionIdLength};
+  quic::QuicCompressedCertsCache compressed_certs_cache_{100};
+  EnvoyQuicTestCryptoServerStreamFactory crypto_stream_factory_;
+
+  Http::MockServerConnectionCallbacks http_connection_callbacks_;
+  NiceMock<Http::MockRequestDecoder> orphan_request_decoder_;
+};
+
+struct Harness {
+  Harness(quic::ParsedQuicVersion quic_version)
+      : quic_version_(quic_version), version_manager(quic::CurrentSupportedHttp3Versions()) {
+    api = Api::createApiForTest();
+    dispatcher = api->allocateDispatcher("envoy_quic_h3_fuzzer_thread");
+    auto connection_helper = std::unique_ptr<quic::QuicConnectionHelperInterface>(
+        new EnvoyQuicConnectionHelper(*dispatcher.get()));
+    auto alarm_factory = std::unique_ptr<quic::QuicAlarmFactory>(
+        new EnvoyQuicAlarmFactory(*dispatcher.get(), *connection_helper->GetClock()));
+    fuzz_dispatcher = std::make_unique<FuzzDispatcher>(quic_version_, &version_manager,
+                                                       std::move(connection_helper),
+                                                       std::move(alarm_factory), *dispatcher.get());
+  }
+  quic::ParsedQuicVersion quic_version_;
+  Api::ApiPtr api;
+  Event::DispatcherPtr dispatcher;
+  std::unique_ptr<FuzzDispatcher> fuzz_dispatcher;
+  quic::QuicVersionManager version_manager;
+};
+
+std::unique_ptr<Harness> harness;
+static void resetHarness() { harness = nullptr; };
+DEFINE_PROTO_FUZZER(const test::common::quic::QuicH3FuzzCase& input) {
+  if (harness == nullptr) {
+    quiche::setVerbosityLogThreshold(0);
+    harness = std::make_unique<Harness>(quic::CurrentSupportedHttp3Versions()[0]);
```

Comment:
Is the harness object persisted across tests? If so, how is tainted state prevention achieved?

Reply:
Yes, the more state I can get away with keeping, the faster the fuzzer runs. There's currently no good way of knowing whether state is being tainted that shouldn't be other than testing two versions, where one definitely does not taint state, because all state is reconstructed on each iteration. Switching to a more conservative fuzzing mode should be straight-forward by re-creating the harness (or the FuzzDispatcher) in each iteration.